### PR TITLE
write melt cookbook

### PR DESCRIPTION
**Pull request by jdannberg**
Add a cookbook that explains how to use melt migration in a model. This partly addresses #948. 
The pull request also includes a visualization postprocessor option for the melt fraction, which is a property computed in models that have melting (if they are a `MeltFractionModel`), but could not be visualized so far.

### CODE REVIEW

**Comment by jdannberg** [comment]
Oh, and of course the cookbook will only run after we have merged #1825.

File: cookbooks/global_melt.prm [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+set Max nonlinear iterations               = 20
+set Nonlinear solver tolerance             = 1e-5
+set Number of cheap Stokes solver steps    = 100
+
+# In addition, melting and freezing normally happens on a much faster
+# time scale than the flow of melt, so we want to decouple the advection
+# of melt (and temperature) and the melting process itself. To do that, 
+# we use the operator splitting scheme, and define that for every 
+# advection time step, we want to do at least 50 reaction time step, 
+# and if these time steps would be larger than 100,000 years, we will do
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: This is a bit confusing, why not just write: "..., and we limit the maximum reaction time step length to ...".

File: cookbooks/global_melt.prm [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+##################### Settings for melt transport ########################
+
+# In models with melt transport, the are additional 'Melt settings' that 
+# let us specify parameters related to the melt transport computations. 
+# One of these is the 'Melt transport threshold': If the porosity is lower
+# than the value specified here, melt will no longer move with a different
+# velocity than the solid, and the system of equations will revert to the
+# usual Stokes equations. Only for porosity larger then the threshold, 
+# the two-phase flow equations apply.
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: What happens for ==0.0 as below? Please add something about that.

File: cookbooks/global_melt.prm [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+# usual Stokes equations. Only for porosity larger then the threshold, 
+# the two-phase flow equations apply. 
+subsection Melt settings
+  set Melt transport threshold                = 0.0
+end
+
+# In models with melt transport, we always need a compositional field with
+# the name 'porosity'. Only the field with that name will be advected with
+# the melt velocity, all other compositional fields will continue to work
+# as before. Material model will typically query for the field with the
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: Material models

File: cookbooks/global_melt.prm [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+subsection Melt settings
+  set Melt transport threshold                = 0.0
+end
+
+# In models with melt transport, we always need a compositional field with
+# the name 'porosity'. Only the field with that name will be advected with
+# the melt velocity, all other compositional fields will continue to work
+# as before. Material model will typically query for the field with the 
+# name porosity to compute all melt material properties. 
+# In addition, the 'melt global' model also requires a field with the name
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: 'melt global' material model

File: cookbooks/global_melt.prm [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+# Models with melt transport also need an additional boundary condition: 
+# the gradient of the fluid pressure at the model boundaries. This boundary 
+# condition indirectly also prescribes boundary conditions for the melt velocity, 
+# as the melt velocity is related to the fluid pressure gradient via Darcy's law.
+# If we choose the fluid pressure gradient = solid density * gravity, melt will
+# flow in and out of the model (even if the solid can not flow out) according to
+# the dynamic fluid pressure in the model. Conversely, if we choose the 
+# fluid pressure gradient = fluid density * gravity, melt will flow in or out
+# with the same velocity as the solid (so for a closer boundary, no melt will
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: closed boundary?

File: cookbooks/global_melt.prm [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+    # a term that scales with one over the porosity. This means that for zero 
+    # porosity, the rock can not dilate/compact any more, which is the same
+    # behaviour that we have for solid mantle convection. 
+    set Reference bulk viscosity          = 1e19
+    
+    # In dependence of how much melt is present, we also weaken the shear
+    # viscosity: The more melt is present, the weaker the rock gets. 
+    # This scaling is exponential, following the relation
+    # viscosity ~ exp(-alpha * DeltaT),
+    # where alpha is the parameter given here, and DeltaT is the deviation from the
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: I think you need to replace DeltaT by the porosity, also in the description

File: cookbooks/global_melt.prm [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+  end
+
+  # refine where the porosity is bigger than 1e-4
+  subsection Composition threshold
+    set Compositional field thresholds = 1e-4,1.0
+  end
+end
+
+# As the model is compressible and has an adiabatic temperature profile, we include 
+# adiabatic heating in the list of heating models.
```
Comment (gassmoeller) [inline cookbooks/global_melt.prm]: Please add a comment about shear heating, and why it is left out

File: cookbooks/global_no_melt.prm [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+# resolution of 4 is sufficient to capture the behaviour of
+# upwellings and downwellings.
+subsection Mesh refinement
+  set Initial adaptive refinement              = 0
+  set Initial global refinement                = 4
+  set Time steps between mesh refinement       = 0
+end
+
+# As the model is compressible and has an adiabatic temperature profile, we include 
+# adiabatic heating in the list of heating models.
```
Comment (gassmoeller) [inline cookbooks/global_no_melt.prm]: Also include the shear heating comment here.

File: doc/manual/manual.tex [hunk 9]
```diff
@@ -8094,6 +8094,49 @@ \subsubsection{Inner core convection}
...
+The following input file (which can be found in \url{cookbooks/global_no_melt.prm}) contains a detailed description of the 
+different options required to set up such a model:
+
+\lstinputlisting[language=prmfile]{cookbooks/global_melt/global_no_melt.prm}
+
+When we look at visualization output of this model, we can see that over time, first upwellings, and the downwellings start to form. Both are more or less stable over time, and only change their positions slowly. As melt does not move relative to the solid, broad stable zones of melting with melt fraction of 10\% or more form in areas where material is upwelling.
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: first upwellings, and then downwellings

File: source/postprocess/visualization/melt_fraction.cc [hunk 10]
```diff
@@ -322,7 +322,10 @@ namespace aspect
                                                   "default of 15\\%. "
                                                   "The corresponding p-T-diagrams can be generated by running "
                                                   "the tests melt\\_postprocessor\\_peridotite and "
-                                                  "melt\\_postprocessor\\_pyroxenite.")
+                                                  "melt\\_postprocessor\\_pyroxenite."
```
Comment (gassmoeller) [inline source/postprocess/visualization/melt_fraction.cc]: Please change this postprocessor to use the material model if possible. This prevents confusion about differently computed melt fractions.

**Comment by jdannberg** [comment]
Okay, I think I addressed your comments. I also added a test for the new melt fraction postprocessor functionality.

**Comment by gassmoeller** [comment]
Ready to merge when the tester is finished.

**Pull request merged**